feat: adjustable rule_group interval

wbollock opened the pull request:
This allows for more expensive SLO recording rules by letting users create a custom Prometheus `rule_group.interval`.

https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#rule_group

Now instead of all rule groups using the default global evaluation interval, a custom interval can be set on all sets of recording rules for an SLO.

If no `interval` is set then the global default will be assumed, matching current Sloth behavior.

Resolves: https://github.com/slok/sloth/issues/367

wbollock committed:
feat: adjustable rule_group interval

This allows for more expensive SLO recording rules by letting users
create a custom Prometheus `rule_group.interval`.

https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#rule_group

Now instead of all rule groups using the default global evaluation
interval, a custom interval can be set on all sets of recording rules
for an SLO.

If no `interval` is set then the global default will be assumed,
matching current Sloth behavior.

Resolves: https://github.com/slok/sloth/issues/367

wbollock committed:
docs: add example file for custom rule groups

wbollock committed:
ref: only append RuleGroupInterval when not empty

Related to how the unit test coverage works, this will make sure
RuleGroupInterval is not appended to RuleGroups unless it is defined and
not empty.

Not requring the `yaml` file does an okay job of this but expliclity
writing the rules without RuleGroupInterval is safer.

It's a bit ugly and repetitive but I think that's just how Go works..

wbollock committed:
test: update tests for custom rule_group interval

This updates tests include `interval`'s and also includes a test to make
sure the rules render correctly when interval is not included.

wbh1 commented on internal/prometheus/model.go:
```diff
@@ -39,16 +39,17 @@ type AlertMeta struct {
+	Name              string `validate:"required,name"`
+	Description       string
+	Service           string `validate:"required,name"`
+	RuleGroupInterval string
```
Why not make this `time.Duration`, too?

wbollock replied:
Originally I did but I had to convert it to a string to parse the duration anyway into the `prommodel` format. I can switch back!

wbh1 reviewed (commented):
Overall, looks good! Since Sloth produces 3 rules groups, I think it'd be nice to support specifying separate intervals for each of them, though. (`sloth-slo-meta-recordings`, `sloth-slo-alerts`, and `sloth-slo-sli-recordings` are the 3 prefixes). For example, you may want an interval of `2m` on the SLI recording rules but still `1m` on the `alerts`.

wbollock committed:
ref: add rule intervals per rule type

Instead of a singular global default, now a rule_group interval can be
set for every individual type of rule_group Sloth generates.

The generic, `interval:all` rule will also stay and can "fill in" any missing per-rule
group defaults. Along with the default behavior of doing nothing if no
`interval` is specified.

wbollock committed:
test: add tests for custom rule group intervals

Specific ones and general/specific combined

wbollock committed:
chore: update example for custom rule groups

wbollock commented:
@wbh1 commits added for per rule group intervals:
a94fee994c770e741571f9d70efeee51890cd5f1
ca78450231e2ded7baa2f995b3ff7f97af933be0
dc3994be548f1598c633cc19de25b8b45a3dc416

wbh1 commented on internal/prometheus/storage.go:
```diff
@@ -49,24 +49,95 @@ func (i IOWriterGroupedRulesYAMLRepo) StoreSLOs(ctx context.Context, slos []Stor
+				var ruleGroupIntervalDuration prommodel.Duration
+				var err error
+
+				// if we have a valid meta rule rule_group interval, use that firs
```
```suggestion
				// if we have a valid meta rule rule_group interval, use that first and overwrite any generic ones
```

wbh1 commented on internal/prometheus/storage.go:
```diff
@@ -49,24 +49,95 @@ func (i IOWriterGroupedRulesYAMLRepo) StoreSLOs(ctx context.Context, slos []Stor
+				var ruleGroupIntervalDuration prommodel.Duration
+				var err error
+
+				// if we have a valid meta rule rule_group interval, use that firs
```
```suggestion
				// if we have a valid meta rule rule_group interval, use that first and overwrite any generic ones
```

wbh1 commented on internal/prometheus/storage.go:
```diff
@@ -49,24 +49,95 @@ func (i IOWriterGroupedRulesYAMLRepo) StoreSLOs(ctx context.Context, slos []Stor
-				Rules: slo.Rules.SLIErrorRecRules,
-			})
+
+			// 0s is default empty string value for time.Duration
+			if slo.SLO.RuleGroupInterval.String() != "0s" || slo.SLO.SLIErrorRulesInterval.String() != "0s" {
+				var ruleGroupIntervalDuration prommodel.Duration
+				var err error
+
+				// if we have a valid meta rule rule_group interval, use that first and overwrite any generic ones
+				if slo.SLO.SLIErrorRulesInterval.String() != "0s" {
+					ruleGroupIntervalDuration, err = prommodel.ParseDuration(slo.SLO.SLIErrorRulesInterval.String())
+				} else {
+					ruleGroupIntervalDuration, err = prommodel.ParseDuration(slo.SLO.RuleGroupInterval.String())
+				}
+
+				if err != nil {
+					return fmt.Errorf("could not parse rule_group interval duration %w", err)
+				}
+
+				ruleGroups.Groups = append(ruleGroups.Groups, ruleGroupYAMLv2{
+					Name:              fmt.Sprintf("sloth-slo-sli-recordings-%s", slo.SLO.ID),
+					RuleGroupInterval: ruleGroupIntervalDuration,
+					Rules:             slo.Rules.SLIErrorRecRules,
+				})
+			} else {
+				ruleGroups.Groups = append(ruleGroups.Groups, ruleGroupYAMLv2{
+					Name:  fmt.Sprintf("sloth-slo-sli-recordings-%s", slo.SLO.ID),
+					Rules: slo.Rules.SLIErrorRecRules,
+				})
+
+			}
```
I think you can simplify this with a `switch` statement like:

```go
    group := ruleGroupYAMLv2{
					Name:             fmt.Sprintf("sloth-slo-sli-recordings-%s", slo.SLO.ID),
					Rules:             slo.Rules.SLIErrorRecRules,
				}
	switch {
	case AlertRulesInterval.String() != "0s":
		ruleGroupIntervalDuration, err = time.ParseDuration(AlertRulesInterval.String())
        if err != nil {
            return fmt.Errorf("could not parse rule_group interval duration for alerts %w", err)
        } else {
            group.RuleGroupInterval = ruleGroupIntervalDuration
        }
	case RuleGroupInterval.String() != "0s":
		ruleGroupIntervalDuration, err = time.ParseDuration(RuleGroupInterval.String())
        if err != nil {
            return fmt.Errorf("could not parse default ("all") rule_group interval duration %w", err)
        } else {
            group.RuleGroupInterval = ruleGroupIntervalDuration
        }
	}

    ruleGroups.Groups = append(ruleGroups.Groups, group)
```

wbh1 reviewed (approved):
Approving as it looks good as-is, but left some comments on form.

wbollock committed:
docs: finish comments

Co-authored-by: Will Hegedus <will@wbhegedus.me>

wbollock committed:
ref: rule interval groups less repetitive

wbollock committed:
docs: clean up custome rule group example
